File: 3219-make-lexicographically-smallest-array-by-swapping-elements/make-lexicographically-smallest-array-by-swapping-elements.java
import java.util.*;

class Solution {
    public int[] lexicographicallySmallestArray(int[] nums, int limit) {
        int n = nums.length;
        int[][] sorted = new int[n][2];
        for (int i = 0; i < n; i++) {
            sorted[i][0] = nums[i];
            sorted[i][1] = i;
        }

        // Sort by value
        Arrays.sort(sorted, (a, b) -> Integer.compare(a[0], b[0]));

        int[] result = new int[n];
        int i = 0;

        while (i < n) {
            int j = i;
            List<Integer> indices = new ArrayList<>();
            
            // Find all elements in the same connected component
            while (j < n && (j == i || sorted[j][0] - sorted[j - 1][0] <= limit)) {
                indices.add(sorted[j][1]);
                j++;
            }

            // Sort original indices to place smallest values in leftmost positions
            Collections.sort(indices);

            // Assign sorted values of the component back to original indices
            for (int k = 0; k < indices.size(); k++) {
                result[indices.get(k)] = sorted[i + k][0];
            }

            i = j; // Move to next group
        }

        return result;
    }
}

File: 3219-make-lexicographically-smallest-array-by-swapping-elements/make-lexicographically-smallest-array-by-swapping-elements.c
/**
 * Note: The returned array must be malloced, assume caller calls free().
 */
#include <stdio.h>
#include <stdlib.h>

typedef struct {
    int val;
    int idx;
} Element;

// Comparator to sort elements by value
int compareElements(const void *a, const void *b) {
    Element *e1 = (Element *)a;
    Element *e2 = (Element *)b;
    return (e1->val > e2->val) - (e1->val < e2->val);
}

// Comparator to sort integer indices
int compareInts(const void *a, const void *b) {
    return (*(int *)a - *(int *)b);
}

int* lexicographicallySmallestArray(int* nums, int numsSize, int limit, int* returnSize) {
    Element* sorted = (Element*)malloc(numsSize * sizeof(Element));
    for (int i = 0; i < numsSize; i++) {
        sorted[i].val = nums[i];
        sorted[i].idx = i;
    }

    // Sort element pairs by value
    qsort(sorted, numsSize, sizeof(Element), compareElements);

    int* result = (int*)malloc(numsSize * sizeof(int));
    int* indices = (int*)malloc(numsSize * sizeof(int));
    
    int i = 0;
    while (i < numsSize) {
        int j = i;
        int groupSize = 0;

        // Collect elements belonging to the same component
        while (j < numsSize && (j == i || sorted[j].val - sorted[j - 1].val <= limit)) {
            indices[groupSize++] = sorted[j].idx;
            j++;
        }

        // Sort indices for the current group
        qsort(indices, groupSize, sizeof(int), compareInts);

        // Map smallest values to smallest indices
        for (int k = 0; k < groupSize; k++) {
            result[indices[k]] = sorted[i + k].val;
        }

        i = j; // Advance to next group
    }

    free(sorted);
    free(indices);

    *returnSize = numsSize;
    return result;
}